lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java

```diff
@@ -165,11 +169,35 @@ public void collect(int doc) throws IOException {
               updateMinCompetitiveScore(scorer);
             }
           }
+          if (canEarlyTerminate) {
+              // When early terminating, stop collecting hits from this leaf once we have its prorated hits.
+              if (leafHits > leafHitsThreshold) {
+                  totalHitsRelation = Relation.GREATER_THAN_OR_EQUAL_TO;
+                  throw new CollectionTerminatedException();
+              }
+          }
         }
 
       };
     }
 
+    /** The total number of documents that matched this query; may be a lower bound in case of early termination. */
+    @Override
+    public int getTotalHits() {
+      return totalHits;
+    }
+
+    private int prorateForSegment(int topK, LeafReaderContext leafCtx) {
+        // prorate number of hits to collect based on proportion of documents in this leaf (segment).
+        // p := probability of a top-k document (or any document) being in this segment
+        double p = (double) leafCtx.reader().numDocs() / leafCtx.parent.reader().numDocs();
```

Comment:
Do we need to guard for divide-by-zero here?

Reply:
Oh yes we do. I guess that would be an empty index. This also made me thing about deleted docs. It would be better to compute this ratio using livedocs I think? Basically we want to use numbers that correspond to what will be collected. Can we easily know the number of live docs in a segment and in the index?

Reply:
Ahh yes we should use `numDocs` -- you can get this from `leafCtx.reader().numDocs()`.

Reply:
I added divide-by-zero check, and I think we were always using `numDocs()`, so this should be good.